File: LeetCode/Medium/2396-strictly-palindromic-number/2396-strictly-palindromic-number.c
int isPalindrome_str(char *str, int start, int end) {
    // 주어진 문자열에서 start, end까지가 palindrome인가?
    // Example
    //      isPalindrome(s, 0, strlen(s)-1) : s 전체가 palindrome인가?
    while (start < end) {
        if (str[start++] != str[end--]) return 0;
    }
    return 1;
}

void reverse(char *str) {
    int len = strlen(str);
    for (int i = 0; i < len / 2; i++) {
        char temp = str[i];
        str[i] = str[len - i - 1];
        str[len - i - 1] = temp;
    }
}

char* intToChar(int n){
    char *str = (char *)malloc(12);
    sprintf(str, "%d", n);
    return str;
}

char * makeBase(int n, int base){
    char * result = (char *)malloc(n);
    int len = 0;
    while (n >= base){
        result[len] = intToChar(n % base);
        len ++;
        n = n/base;
    }

    result[len] = intToChar(n);
    result[len+1] = '\0';

    reverse(result);

    return result;
}

bool isStrictlyPalindromic(int n) {
    for (int i = 2; i < n-1 ; i++){
        char * num = makeBase(n, i);
        if (!isPalindrome_str(num, 0, strlen(num)-1)) return false;
    }
    return true;
}

File: LeetCode/Medium/2396-strictly-palindromic-number/2396-strictly-palindromic-number.py
class Solution:
    def makeBase(self, n:int, base:int):
        result = ""

        while n >= base:
            result = str(n % base) + result
            
            n = n // base
        
        result = str(n) + result

        return result

    def isPalindrome(self, s) -> bool:
        """
        주어진 문자열 혹은 int가 Palindrome인가?
        """

        if isinstance(s, str):
            return s == s[::-1]
        
        elif isinstance(s, int):
            reverse = 0
            temp = s 
            while (temp != 0): 
                reverse = (reverse * 10) + (temp % 10) 
                temp = temp // 10
            
        return (reverse == s) # if it is true then it will return 1; 
    
    def isStrictlyPalindromic(self, n: int) -> bool:
        for i in range(2, n-1):
            num = self.makeBase(n, i)
            if not self.isPalindrome(num): return False
        return True
